### ISSUE TITLE
Fingerprinting slow on linux

### ISSUE THREAD

**Issue by doniwinata0309** [gradle/gradle#28236]
### Current Behavior

I use linux CI to run our project (pretty large with 300+ module) and fingerprinting takes around 20m or above for full build.

Some up-to-date task taking 40-60 sec/each to complete. When using gradle benchmark to try compile avoidance (when most task should be up-to-date) 1st and 2nd build fingerprinting super slow, but 3rd and 4th build getting faster.

If im comparing this to our mac OS build, it is only takes 5minutes for fingerprinting on full build, and each task mostly only takes less than 5 sec.

### Expected Behavior

Fingerprinting for up-to-date task should fast (< 5sec) on linux, similar to mac os.

### Context (optional)

We are changing our CI form mac os to linux and build time increased significantly because fingerprinting issue.

Linux details:
![image](https://github.com/gradle/gradle/assets/38415463/c2ede253-d28b-4592-a45d-b2f4f8ee426f)

Mac OS details:
![image](https://github.com/gradle/gradle/assets/38415463/74d80b95-934c-43a6-a737-fabc211524df)

Do you have any advice how to debug this issue?

### Steps to Reproduce

Unable to repro this on small project.
On our project, we run gradle profiler by changing private method on some modules to trigger compile avoidance. Compile avoidance happen. 
Image below showing 4 iteration, the 2nd and 
![image](https://github.com/gradle/gradle/assets/38415463/f37fcff7-cd11-4e68-8902-5075cd1f8198)
Task execution detail:
![image](https://github.com/gradle/gradle/assets/38415463/4c874d58-aeb2-468c-b1f3-7997eb6c04c1)

Comparing to mac os with same source code:
![image](https://github.com/gradle/gradle/assets/38415463/69e1d1cd-9c2f-450a-a819-251845464b93)

### Gradle version

8.5

### Gradle version that used to work

-

**Comment by ov7a** [comment]
Thank you for your interest in Gradle!

This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

**Comment by lptr** [comment]
@oehme could your changes be relevant here?

**Comment by oehme** [comment]
The issues I've found so far are platform-independent, but any fingerprinting issues are interesting to me. @doniwinata0309 since you are already using gradle-profiler, could you share a profile with the slow fingerprinting time, preferably using the `async-profiler` option?

**Comment by doniwinata0309** [comment]
Please find attachment for async profiler and chrome trace:
[async_profiler.zip](https://github.com/gradle/gradle/files/14416109/async_profiler.zip)
[compile_avoidance_app-8.5-trace.json.zip](https://github.com/gradle/gradle/files/14416461/compile_avoidance_app-8.5-trace.json.zip)
fingerprinting for this build is around 34m
![image](https://github.com/gradle/gradle/assets/38415463/e6daebf1-0818-4b71-9e71-b4954a037b28)

Appreciate your help

**Comment by doniwinata0309** [comment]
[linux.zip](https://github.com/gradle/gradle/files/14421775/linux.zip)
This is the result on linux. for mac not sure why it is become very large like 500mb for 1 stack file. will retry again and send youthe result later

**Comment by oehme** [comment]
From what I'm seeing in that snapshot, it looks like everything is waiting on the file hash cache, which is a small database on disk that gets lots of tiny reads and writes during a build. What kind of machine are you running this on? For example, on AWS, I'd strongly recommend agents with SSDs instead of EBS volumes.

**Comment by doniwinata0309** [comment]
It is
c7i.4xlarge | Processor: 4th Generation Intel Xeon Scalable | vCPU: 16 | Memory: 32 | EBS-Only | Network badnwidth: Up to 12.5 | EBS bandwith: Up to 10 

I will try to benchmark with 
c6id.4xlarge with SSD and update you the result.

Thank you, appreciate your help

**Comment by doniwinata0309** [comment]
also wondering, is there any other workaround i can do to optimize it on that machine?

**Comment by oehme** [comment]
I have a [possible fix](https://github.com/gradle/gradle/compare/master...oehme:gradle:oehme/faster-btree-cache) for you to try out, if you're up for building Gradle from source. You'd have to check out that branch of my fork, run `gradlew install -Pgradle_installPath=<where to put the installation>` and then run your experiments with that custom installation.

**Comment by doniwinata0309** [comment]
Sure let me try that later, Thank you

**Comment by oehme** [comment]
> The result on AWS using SSD is much faster compared to EBS, and we will consider to use ssd for our CI. 

That is my experience as well, glad you're seeing a good speedup.

> But sometimes i can see task like com.android.build.gradle.internal.tasks.DexMergingTask takes 40 sec to just fingerprinting, executing 0.6sec. Is there a way to skip fingerprinting on some task, forcing it to just run without cache/checking up-to-date?

Actually, this is misleading - DexMerging depends on artifact transformations, which are currently counted towards "fingerprinting", but it's actually real work being done by the Android plugin. So disabling fingerprinting in that case won't do anything.

> unfortunately i got some error after using this [possible fix](https://github.com/gradle/gradle/compare/master...oehme:gradle:oehme/faster-btree-cache)
Gradle dissapear, probably there some issue with our app when using gradle 8.8.

I'd love to know more - Have you tried with the `master` branch for comparison?

**Comment by doniwinata0309** [comment]
Ah i see so that may be artifact transform causing issue. I will try later with gradle 8.5 source code and make changes similar to your commit. Will update the result later, thank you for your help.

**Comment by doniwinata0309** [comment]
This is the result from 2 benchmark with 3 iteration (1 warm up)
Gradle 8.5 with [possible fix](https://github.com/gradle/gradle/compare/master...oehme:gradle:oehme/faster-btree-cache)
![image](https://github.com/gradle/gradle/assets/38415463/9bb717a0-4239-426e-89e7-11e6d0ef3260)

Gradle 8.5 and 8.2
![image](https://github.com/gradle/gradle/assets/38415463/8924ef07-821c-4865-8783-57bf7371c800)

Looks like it can fix the issue. 

To use this on our CI, do i need to copy source --> run build --> point out gradle to use custom installtion path ?
Or do you have advice how we can use custom gradle temporarily like use .jar or something?

**Comment by sorin-florea** [comment]
@doniwinata0309 What version of the Develocity plugin are you using? Using the Develocity Gradle plugin 3.16+ with Gradle 8.3+ will accurately display the transform execution time and fingerprinting time separately.

Develocity plugin 3.15
![Screenshot 2024-03-04 at 2 56 08 PM](https://github.com/gradle/gradle/assets/30589784/d5e900be-af0d-4f6c-8550-2cd0f56ac93c)

Devlocity plugin 3.16
![Screenshot 2024-03-04 at 2 56 17 PM](https://github.com/gradle/gradle/assets/30589784/04183b1c-3120-4fae-8d17-b88a72a8fe80)

**Comment by doniwinata0309** [comment]
@sorin-florea it is Develocity plugin 3.15.1, i will update it to 3.16 thanks

**Comment by lptr** [comment]
It looks like @oehme's changes would introduce some improvements in performance on slow disks. However, we have some plans to replace this infrastructure in the foreseeable future completely. See also notes here: https://github.com/gradle/gradle/pull/28344#issuecomment-2018516744

**Comment by lptr** [comment]
Also, as @oehme recommended above, switching to SSDs could already improve performance significantly.

**Comment by bigdaz** [comment]
I suspect the same issue is leading to very slow `JavaCompile` tasks when incremental compile is enabled: https://github.com/gradle/dv/issues/62253#issuecomment-4017088571

I updated PR #28344 and applied it to master, then ran a bunch of experiments comparing Gradle 9.3.1 with the patched version. ([GHActions run](https://github.com/bigdaz/incremental-compile-test/actions/runs/22984214012))

Here are results with no incremental compile data in Gradle User Home, incremental compile enabled. All tests were run on the exact same hardware: the "Fast disk" is instance-storage (local disk) and "Slow disk" is the default EBS disk allocated to the EC2 instance (with only 3000 IOPs provisioned).

Fast disk
Gradle 9.3.1: [155s](https://ge.solutions-team.gradle.com/s/vmbnno36cmdvq)
Patched Gradle: [77s](https://ge.solutions-team.gradle.com/s/tyesity4azhde)

Slow disk:
Gradle 9.3.1: [649s](https://ge.solutions-team.gradle.com/s/shnqxycu3hzvy)
Patched Gradle: [78s](https://ge.solutions-team.gradle.com/s/d52rjyaosnqay) (8x faster!)


Results with incremental compile data restored, incremental compile enabled.
This is the scenario with setup-gradle, which does restore javaCompile state.

Fast disk
Gradle 9.3.1: [101s](https://ge.solutions-team.gradle.com/s/ide3l2nvk322k)
Patched Gradle: [79s](https://ge.solutions-team.gradle.com/s/mkcsojnncuigw)

Slow disk:
Gradle 9.3.1: [133s](https://ge.solutions-team.gradle.com/s/faqvv4qlrxmzu)
Patched Gradle: [95s](https://ge.solutions-team.gradle.com/s/7n5wcvqayaxxi)


Finally with incremental compile disabled. Restoring javaCompile should make no difference.

Fast disk
Gradle 9.3.1: [133s](https://ge.solutions-team.gradle.com/s/y6geqmrusomps)
Patched Gradle: [77s](https://ge.solutions-team.gradle.com/s/nie677vhkipie)

Slow disk:
Gradle 9.3.1: [169s](https://ge.solutions-team.gradle.com/s/3er4hnjhkn7mg)
Patched Gradle: [77s](https://ge.solutions-team.gradle.com/s/zpbivs7sjb246)


As you can see, while the slow-disk + incremental compile is the most notable, these are very significant improvements in all scenarios.

The underlying issue is that FileBackedBlockStore writes to the physical disk on every change, even if there's not contention. Stefan's fix was to buffer in memory (if we have an exclusive lock) and only write when complete. When the disk has a limited number of operations per second, this fix removes a critical blocker. The fix is actually pretty simple: https://github.com/gradle/gradle/pull/28344/changes/52c3ada22f7392f337e9770d7d15c9416a66b619 (edited)

**Comment by asodja** [comment]
Note: A fix that should improve this scenario was done in https://github.com/gradle/gradle/pull/37386

**Comment by ljacomet** [comment]
Given the fix going in Gradle 9.6.0, marking this as fixed.

Anyone having the issue on older versions should either upgrade Gradle or upgrade the IOPs of their storage as a workaround.
